### PR TITLE
Refactor PointerLockControls

### PR DESCRIPTION
**Pull request by WestLangley**
This is reasonable attempt at implementing #16160.

/ping @arodic 
/ping @Mugen87

### CODE REVIEW

File: examples/js/controls/PointerLockControls.js [hunk 1]
```diff
@@ -10,14 +10,7 @@ THREE.PointerLockControls = function ( camera, domElement ) {
...
 
-	camera.rotation.set( 0, 0, 0 );
-
-	var pitchObject = new THREE.Object3D();
-	pitchObject.add( camera );
-
-	var yawObject = new THREE.Object3D();
-	yawObject.position.y = 10;
-	yawObject.add( pitchObject );
+	camera.rotation.reorder( 'YXZ' );
```
Comment (Mugen87) [inline examples/js/controls/PointerLockControls.js]: Could this have a side effect? What if the user disconnects the controls at some point and then changes the rotation manually? The result will be unexpected if the euler order is not the default. 

Could also be problematic if `PointerLockControls` is used with other controls.

**Comment by arodic** [comment]
There is no need to change rotation order on the camera.

Pointerlock behavior can be calculated with two quaternions then applied to camera rotation in correct order.

File: examples/js/controls/PointerLockControls.js (hunk 1 above)
Comment (WestLangley) [reply examples/js/controls/PointerLockControls.js]: >Could this have a side effect? What if the user disconnects the controls at some point and then changes the rotation manually?

You can change it back upon disconnect, then.

>Could also be problematic if PointerLockControls is used with other controls, right?

Can you please provide an example?

**Comment by WestLangley** [comment]
>There is no need to change rotation order on the camera.

This is exactly what the `rotation.order` is for. If you want the Euler rotation to represent _heading_, _pitch_, and _roll_, then you set `rotation.order` accordingly.

In fact, we should have made `YXZ` the default order in the first place. It would have resulted in much less user confusion. Too late now, unfortunately...

>Pointerlock behavior can be calculated with two quaternions then applied to camera rotation in correct order.

Can you share your implementation of that?

File: examples/js/controls/PointerLockControls.js (hunk 1 above)
Comment (Mugen87) [reply examples/js/controls/PointerLockControls.js]: I've thought of a custom control class that also changes the rotation via euler angles. In this case, you have to be aware that `PointerLockControls` changes the euler order and act accordingly.

Comment (WestLangley) [reply examples/js/controls/PointerLockControls.js]: You are referring to hypotheticals.

**Comment by arodic** [comment]
> Can you share your implementation of that?

Sure. Probably tomorrow.

Ideally, users should not require knowledge of control internals to use it. If control changes camera rotation order, it shouldn't be in in lock() function (not the constructor), and it should restore it to previous order in unlock().

> You are referring to hypotheticals.

It's really not that difficult to imagine how changing rotation order can mess up camera behavior. I could come up with a few examples, or even create live examples but honestly it would be a waste of time. PointerLockControls is a very simple utility and it's trivial to implement it without mutating default camera parameters and without requiring users to reset it back to default.

**Comment by WestLangley** [comment]
>I could come up with a few examples, or even create live examples but honestly it would be a waste of time. 

Closing then.

File: examples/js/controls/PointerLockControls.js (hunk 1 above)
Comment (Mugen87) [reply examples/js/controls/PointerLockControls.js]: I just think it's a bad design decision to change the euler order in controls.

Comment (WestLangley) [reply examples/js/controls/PointerLockControls.js]: That is an opinion and a preference.  That is fine. 

But to file objections without specifics that the OP can respond to is unfair to the OP -- me. It also has the effect of blocking a PR from ever being merged.

And as I said, the order can be restored upon disconnect -- or at the end of every update if you insist.

**Comment by WestLangley** [comment]
Reopened and refactored in response to feedback.

**Comment by mrdoob** [comment]
@arodic I'm thinking on merging this PR and maybe you can build yours on top of this one. What do you think?

**Comment by arodic** [comment]
@mrdoob what features would make sense to merge?

I think the most crucial is the **change event**. It should be emitted on mousemove, lock and unlock. This enables controls to work with applications without continuous render loop such as the editor.

Reverse vertical movement parameter is a nice to have and can be easily applied to this PR.

Min/max polar angle and polar/azimuth angle getters, I think should rely on spherical for clarity and consistency with other controls. Therefore it's not suitable for this implementation in my opinion

**Comment by arodic** [comment]
@mrdoob of you merge this, I can fix some of the other issues on top.

**Comment by WestLangley** [comment]
Since @mrdoob has agreed to this, merging now so @arodic can add additional features.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**